src/darwin/Framework/CHIPTests/MTRDeviceTests.m

```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+{
+    XCTAssertNotNil(endpointId);
+    XCTAssertNotNil(device);
+    XCTAssertNotNil(delegate);
+    XCTAssertNotNil(dataVersion);
+    XCTAssertNotNil(attributeReport);
+    XCTAssertNotNil(testcase);
+
+    XCTestExpectation * gotAttributeReportExpectation = [testcase expectationWithDescription:@"Attribute report has been received"];
+    XCTestExpectation * gotAttributeReportEndExpectation = [testcase expectationWithDescription:@"Attribute report has ended"];
+    XCTestExpectation * deviceConfigurationChangedExpectation = [testcase expectationWithDescription:@"Device configuration changed was receieved"];
+    deviceConfigurationChangedExpectation.inverted = !expectConfigurationChanged;
+
+    __block unsigned attributeReportsReceived = 0;
+    __block NSArray<NSDictionary<NSString *, id> *> * testDataValue = nil;
+
+    for (NSDictionary<NSString *, id> * attributeDict in attributeReport) {
+        MTRAttributePath * attributePath = attributeDict[MTRAttributePathKey];
+        if (attributePath.attribute.unsignedLongLongValue == attributeId) {
+            NSDictionary * data = attributeDict[MTRDataKey];
+            XCTAssertNotNil(data);
+            testDataValue = data[MTRValueKey];
+        }
+    }
+
+    // Check if the received attribute report matches the injected attribute report.
+    delegate.onAttributeDataReceived = ^(NSArray<NSDictionary<NSString *, id> *> * attributeReport) {
+        attributeReportsReceived += attributeReport.count;
+        XCTAssert(attributeReportsReceived > 0);
+        for (NSDictionary<NSString *, id> * attributeDict in attributeReport) {
+            MTRAttributePath * attributePath = attributeDict[MTRAttributePathKey];
+            XCTAssertNotNil(attributePath);
+
+            XCTAssertEqualObjects(attributePath.cluster, @(clusterId));
+            XCTAssertEqualObjects(attributePath.attribute, @(attributeId));
+
+            NSDictionary * data = attributeDict[MTRDataKey];
+            XCTAssertNotNil(data);
+            XCTAssertEqualObjects(data[MTRDataVersionKey], dataVersion);
+
+            NSArray<NSNumber *> * dataValue = data[MTRValueKey];
+            XCTAssertNotNil(dataValue);
+            XCTAssertNotNil(testDataValue);
+            XCTAssertEqualObjects(dataValue, testDataValue);
+            [gotAttributeReportExpectation fulfill];
+        }
+    };
+
+    delegate.onReportEnd = ^() {
+        [gotAttributeReportEndExpectation fulfill];
+    };
+
+    __block BOOL wasOnDeviceConfigurationChangedCallbackCalled = NO;
+
+    delegate.onDeviceConfigurationChanged = ^() {
+        [deviceConfigurationChangedExpectation fulfill];
+        wasOnDeviceConfigurationChangedCallbackCalled = YES;
+    };
+
+    [device unitTestInjectAttributeReport:attributeReport];
+
+    [testcase waitForExpectations:@[ gotAttributeReportExpectation, gotAttributeReportEndExpectation, deviceConfigurationChangedExpectation ] timeout:kTimeoutInSeconds];
```

Comment:
Huh, so waiting on an inverted expectation works fine?  Good to know!

Reply:
yup I was going to ask you yesterday if thats ok and found that it is during testing. good to know